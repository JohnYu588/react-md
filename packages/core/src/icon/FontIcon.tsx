import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";

import type { FontIconClassNameOptions } from "./styles";
import { icon } from "./styles";

export interface FontIconProps
  extends Omit<HTMLAttributes<HTMLElement>, "color">,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 *
 * \@remarks \@since 6.0.0 Switched from `<i>` to `<span>` element.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(props, ref) {
    const {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      color,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...remaining
    } = props;

    return (
      <span
        {...remaining}
        aria-hidden={ariaHidden}
        ref={ref}
        className={icon({
          type: "font",
          dense,
          color,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </span>
    );
  }
);
