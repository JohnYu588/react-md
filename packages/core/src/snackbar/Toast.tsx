"use client";

import {
  type AriaRole,
  type HTMLAttributes,
  type ReactElement,
  type ReactNode,
  forwardRef,
  isValidElement,
} from "react";

import { type ButtonProps } from "../button/Button.js";
import { type BackgroundColor } from "../cssUtils.js";
import {
  type CSSTransitionClassNames,
  type TransitionCallbacks,
  type TransitionTimeout,
} from "../transition/types.js";
import { useScaleTransition } from "../transition/useScaleTransition.js";
import { type PropsWithRef } from "../types.js";
import { useEnsuredId } from "../useEnsuredId.js";
import { ToastActionButton } from "./ToastActionButton.js";
import { ToastCloseButton } from "./ToastCloseButton.js";
import { ToastContent, type ToastContentProps } from "./ToastContent.js";
import { toast } from "./toastStyles.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-toast-color"?: string;
    "--rmd-toast-background-color"?: string;
    "--rmd-toast-offset"?: string | number;
  }
}

/**
 * @since 6.0.0
 */
export interface ConfigurableToastProps
  extends HTMLAttributes<HTMLDivElement>,
    TransitionCallbacks {
  /**
   * Note: this default value will only be generated in the `Toast` component.
   *
   * @defaultValue `"toast-" + useId()`
   */
  id?: string;

  /**
   * Note: This is set while creating the toast.
   *
   * @defaultValue `visibleTime === null ? "alert" : "status"`
   */
  role?: AriaRole;

  /**
   * @defaultValue `"surface"`
   */
  theme?: BackgroundColor;

  /**
   * Set this to `true` to stack the content above the {@link action}. It is not
   * recommended to enable this prop if the {@link closeButton} is enabled.
   *
   * @defaultValue `false`
   */
  stacked?: boolean;

  /**
   * If this is not provided, a `ResizeObserver` will be used to determine if
   * there are multiple lines of content.
   */
  multiline?: boolean;

  /**
   * When this is a string or React element, it will be rendered as the
   * `children` within a `Button`
   */
  action?: ButtonProps | ReactElement | string;

  /**
   * This can be used to replace the custom action button behavior.
   */
  actionButton?: ReactNode;

  /**
   * @defaultValue `getIcon("close")`
   */
  closeIcon?: ReactNode;

  /**
   * Set this to `true` if a close button should be rendered to the right of the
   * `children`.
   *
   * @defaultValue `!!closeButtonProps`
   */
  closeButton?: boolean;

  /**
   * Use this prop to override most of the close button behavior. The
   */
  closeButtonProps?: ButtonProps;

  /**
   * Any additional props that should be provided to the `<div>` that surrounds
   * the toast `children`.
   */
  contentProps?: PropsWithRef<ToastContentProps, HTMLDivElement>;

  /**
   * Set this to `true` if the `children` for the toast should no longer be
   * wrapped in an additional `<div>` that applies some toast layout styles.
   * This should normally only be used for custom `Toast` implementations.
   *
   * @see the `Snackbar`'s `renderToast` prop for an example.
   */
  disableToastContent?: boolean;

  /**
   * The toast's transition timeout for entering and exiting. This is **not**
   * how long the toast should remain visible.
   *
   * @defaultValue `SCALE_TIMEOUT`
   */
  timeout?: TransitionTimeout;

  /**
   * The toast's transition class names for entering and exiting.
   *
   * @defaultValue `SCALE_CLASSNAMES`
   */
  classNames?: CSSTransitionClassNames;
}

/**
 * @since 6.0.0
 */
export interface ToastProps extends ConfigurableToastProps {
  paused?: boolean;
  visible: boolean;
}

/**
 * **Client Component**
 *
 * This component is just used for toast styling and does not implement any of
 * the visibility behavior.
 *
 * @see {@link https://next.react-md.dev/components/snackbar | Snackbar Demos}
 * @since 6.0.0
 */
export const Toast = forwardRef<HTMLDivElement, ToastProps>(
  function Toast(props, ref) {
    const {
      id: propId,
      className,
      timeout,
      classNames,
      theme = "surface",
      action: propAction,
      actionButton: propActionButton,
      paused,
      visible,
      closeIcon: propCloseIcon,
      closeButtonProps,
      closeButton = !!closeButtonProps,
      contentProps,
      disableToastContent,
      stacked,
      multiline,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      children,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "toast");

    let actionButton = propActionButton;
    if (propAction) {
      let overrides: ButtonProps = {};
      let buttonChildren: ReactNode;
      // have to use `any` to correctly filter out all react elements
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      if (isValidElement<any>(propAction) || typeof propAction !== "object") {
        buttonChildren = propAction;
      } else {
        ({ children: buttonChildren, ...overrides } = propAction);
      }

      actionButton = (
        <ToastActionButton
          theme={theme === "surface" ? "secondary" : "clear"}
          reordered={stacked && closeButton}
          {...overrides}
        >
          {buttonChildren}
        </ToastActionButton>
      );
    }

    let closeIcon = propCloseIcon;
    if (typeof closeButtonProps?.children !== "undefined") {
      closeIcon = closeButtonProps.children;
    }

    const action = !!actionButton;
    const reordered = stacked && action && closeButton;
    const { elementProps, rendered } = useScaleTransition({
      appear: true,
      nodeRef: ref,
      className: toast({
        className,
        theme,
        action,
        paused,
        stacked,
        reordered,
        closeButton,
      }),
      timeout,
      classNames,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      temporary: true,
      transitionIn: visible,
      exitedHidden: true,
    });

    // this might get rid of the weird popping-back-in for a split second
    // that sometimes happens on mobile firefox
    if (!rendered) {
      return null;
    }

    return (
      <div {...remaining} {...elementProps} id={id}>
        <ToastContent
          action={action}
          stacked={stacked}
          multiline={multiline}
          closeButton={closeButton}
          disableWrapper={disableToastContent}
          {...contentProps}
        >
          {children}
        </ToastContent>
        {actionButton}
        {closeButton && (
          <ToastCloseButton reordered={reordered} {...closeButtonProps}>
            {closeIcon}
          </ToastCloseButton>
        )}
      </div>
    );
  }
);
