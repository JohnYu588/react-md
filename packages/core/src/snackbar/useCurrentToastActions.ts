"use client";
import { createContext, useContext } from "react";

/**
 * A simplified version of the {@link ToastManager} that has the actions bound
 * to the current {@link QueuedToast.toastId}.
 *
 * @remarks \@since 6.0.0
 */
export interface CurrentToastActions {
  /** @see {@link ToastManager.clearTimer} */
  clearTimer(): void;
  /** @see {@link ToastManager.removeToast} */
  removeToast(transition: boolean): void;
  /** @see {@link ToastManager.startRemoveTimeout} */
  startRemoveTimeout(): void;
  /** @see {@link ToastManager.pauseRemoveTimeout} */
  pauseRemoveTimeout(): void;
  /** @see {@link ToastManager.resumeRemoveTimeout} */
  resumeRemoveTimeout(): void;
}

const context = createContext<CurrentToastActions | null>(null);
context.displayName = "CurrentToastActions";

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export const { Provider: CurrentToastActionsProvider } = context;

/**
 * This is only required if you have multiple `Snackbar` implementations in your
 * app.
 *
 * @remarks \@since 6.0.0
 */
export function useCurrentToastActions(): CurrentToastActions {
  const actions = useContext(context);
  if (!actions) {
    throw new Error(
      "The `CurrentToastActionsProvider` has not been initialized"
    );
  }

  return actions;
}
