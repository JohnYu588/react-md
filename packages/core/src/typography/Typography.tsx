import {
  forwardRef,
  type ElementType,
  type HTMLAttributes,
  type ReactElement,
} from "react";
import {
  typography,
  type TypographyClassNameOptions,
  type TypographyType,
} from "./typographyStyles.js";

/**
 * A union of the default supported elements that the `Typography` component can
 * be rendered as. This is mostly used for adding the correct `HTMLAttributes`
 * and enabling the forward ref.
 *
 * @since 4.0.0
 */
export type TypographyHTMLElement =
  | HTMLHeadingElement
  | HTMLParagraphElement
  | HTMLSpanElement
  | HTMLDivElement
  | HTMLAnchorElement
  | HTMLBodyElement
  | HTMLHtmlElement;

/** @since 6.0.0 */
export type CustomTypographyComponent = ElementType<
  HTMLAttributes<TypographyHTMLElement> & { className: string }
>;

/** @internal */
function getComponent(
  as: CustomTypographyComponent | undefined,
  type: TypographyType
): ElementType {
  if (as) {
    return as;
  }

  switch (type) {
    case "headline-1":
      return "h1";
    case "headline-2":
      return "h2";
    case "headline-3":
      return "h3";
    case "headline-4":
      return "h4";
    case "headline-5":
      return "h5";
    case "headline-6":
    case "subtitle-1":
    case "subtitle-2":
      return "h6";
    case "body-1":
    case "body-2":
      return "p";
    case "caption":
      return "caption";
    default:
      return "span";
  }
}

export interface TypographyProps
  extends HTMLAttributes<TypographyHTMLElement>,
    TypographyClassNameOptions {
  /**
   * The component to render as when the children are not a render function. If
   * this prop is omitted, the component will be determined by the `type` prop
   * where:
   *
   * - `"headline-1" -> <h1>`
   * - `"headline-2" -> <h2>`
   * - `"headline-3" -> <h3>`
   * - `"headline-4" -> <h4>`
   * - `"headline-5" -> <h5>`
   * - `"headline-6" -> <h6>`
   * - `"subtitle-1" -> <h5>`
   * - `"subtitle-2" -> <h6>`
   * - `"body-1"     -> <p>`
   * - `"body-2"     -> <p>`
   * - `"caption"    -> <caption>`
   * - `"overline"   -> <span>`
   */
  as?: CustomTypographyComponent;
}

/**
 * **Server Component**
 *
 * Render text with one of the material design typography styles applied and
 * optional styles like font-weight, font-style, text color, etc.
 *
 * @example All Example
 * ```tsx
 * import { Typography } from "@react-md/core":
 *
 * export function Example() {
 *   return (
 *    <>
 *      <Typography type="headline-1">Headline 1</Typography>
 *      <Typography type="headline-2">Headline 2</Typography>
 *      <Typography type="headline-3">Headline 3</Typography>
 *      <Typography type="headline-4">Headline 4</Typography>
 *      <Typography type="headline-5">Headline 5</Typography>
 *      <Typography type="headline-6">Headline 6</Typography>
 *      <Typography type="subtitle-1">Subtitle 1</Typography>
 *      <Typography type="subtitle-2">Subtitle 2</Typography>
 *      <Typography>
 *        A paragraph of text.
 *      </Typography>
 *      <Typography type="body-1">
 *        A paragraph of text.
 *      </Typography>
 *      <Typography type="body-2">
 *        Another paragraph of text.
 *      </Typography>
 *      <Typography type="caption" component="h5">
 *        Caption text
 *      </Typography>
 *      <Typography type="overline" component="h5">
 *        Overline text
 *      </Typography>
 *    </>
 *   ):
 * }
 * ```
 */
export const Typography = forwardRef<TypographyHTMLElement, TypographyProps>(
  function Typography(props, ref): ReactElement {
    const {
      as,
      type = "body-1",
      className,
      margin,
      fontStyle,
      fontWeight,
      textAlign,
      textColor,
      textDecoration,
      textTransform,
      textOverflow,
      children,
      ...remaining
    } = props;

    const Component = getComponent(as, type);
    return (
      <Component
        {...remaining}
        ref={ref}
        className={typography({
          type,
          margin,
          fontStyle,
          fontWeight,
          textAlign,
          textColor,
          textDecoration,
          textTransform,
          textOverflow,
          className,
        })}
      >
        {children}
      </Component>
    );
  }
);
