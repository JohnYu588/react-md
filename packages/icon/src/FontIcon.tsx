import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";
import type { FontIconClassNameOptions } from "./getIconClassName";
import { getIconClassName } from "./getIconClassName";

export interface FontIconProps
  extends HTMLAttributes<HTMLElement>,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(
    {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...props
    },
    ref
  ) {
    return (
      <i
        {...props}
        aria-hidden={ariaHidden}
        ref={ref}
        className={getIconClassName({
          type: "font",
          dense,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </i>
    );
  }
);
