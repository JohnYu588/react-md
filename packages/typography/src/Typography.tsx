import type {
  ElementType,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from "react";
import { forwardRef } from "react";
import type { ClassNameCloneableChild } from "@react-md/utils";

import type { TypographyClassNameOptions, TypographyType } from "./styles";
import { typographyClassName } from "./styles";

/**
 * A union of the default supported elements that the `Typography` component can
 * be rendered as. This is mostly used for adding the correct `HTMLAttributes`
 * and enabling the forward ref.
 *
 * @remarks \@since 4.0.0
 */
export type TypographyHTMLElement =
  | HTMLHeadingElement
  | HTMLParagraphElement
  | HTMLSpanElement
  | HTMLDivElement
  | HTMLButtonElement
  | HTMLAnchorElement
  | HTMLBodyElement
  | HTMLHtmlElement;

/** @remarks \@since 4.0.0 */
export type TypographyRenderFunction = (props: {
  className: string;
}) => ReactElement;

/**
 * @remarks \@since REPLACE_VERSION Extends the {@link TypographyClassNameOptions}
 * interface and added missing default value annotations.
 */
export interface TypographyProps
  extends Omit<HTMLAttributes<TypographyHTMLElement>, "color">,
    TypographyClassNameOptions {
  /**
   * An optional className to merge into typography styles.
   */
  className?: string;

  /**
   * The component to render as when the children are not a render function. If
   * this prop is omitted, the component will be determined by the `type` prop
   * where:
   *
   * - `"headline-1" -> <h1>`
   * - `"headline-2" -> <h2>`
   * - `"headline-3" -> <h3>`
   * - `"headline-4" -> <h4>`
   * - `"headline-5" -> <h5>`
   * - `"headline-6" -> <h6>`
   * - `"subtitle-1" -> <h5>`
   * - `"subtitle-2" -> <h6>`
   * - `"body-1"     -> <p>`
   * - `"body-2"     -> <p>`
   * - `"caption"    -> <caption>`
   * - `"overline"   -> <span>`
   * - `"button"     -> <button>`
   *
   * @defaultValue `null`
   */
  component?: ElementType | null;

  /**
   * Either a child render function or a react node. If this is not the child
   * render function, a different wrapper component can be provided using the
   * `component` prop.
   */
  children?: ReactNode | ClassNameCloneableChild | TypographyRenderFunction;
}

function getComponent(
  component: ElementType | null,
  type: TypographyType
): ElementType {
  if (component) {
    return component;
  }

  switch (type) {
    case "headline-1":
      return "h1";
    case "headline-2":
      return "h2";
    case "headline-3":
      return "h3";
    case "headline-4":
      return "h4";
    case "headline-5":
      return "h5";
    case "headline-6":
    case "subtitle-1":
    case "subtitle-2":
      return "h6";
    case "body-1":
    case "body-2":
      return "p";
    case "caption":
      return "caption";
    case "button":
      return "button";
    default:
      return "span";
  }
}

/**
 * The `Typography` component is used to render text with the material design
 * typography styles applied.  By default, everything will be rendered in a
 * `<p>` tag with the normal paragraph styles.
 *
 * When the `type` prop is changed to another typography style, this component
 * will determine the "best" element to render the text in *unless* the
 * `component` prop is provided. The default mapping is:
 *
 * - `"headline-1" -> <h1>`
 * - `"headline-2" -> <h2>`
 * - `"headline-3" -> <h3>`
 * - `"headline-4" -> <h4>`
 * - `"headline-5" -> <h5>`
 * - `"headline-6" -> <h6>`
 * - `"subtitle-1" -> <h5>`
 * - `"subtitle-2" -> <h6>`
 * - `"body-1"     -> <p>`
 * - `"body-2"     -> <p>`
 * - `"caption"    -> <caption>`
 * - `"overline"   -> <span>`
 * - `"button"     -> <button>`
 *
 * NOTE: if the `component` prop is not `null`, this logic will be ignored and
 * the provided `component` will be used instead.
 *
 * @example
 * Simple Example
 * ```tsx
 * import type { ReactElement } from "react";
 * import { Typography } from "@react-md/typography";
 *
 * function Example(): ReactElement {
 *   return (
 *     <>
 *       <Typography type="headline-1">Headline 1</Typography>
 *       <Typography type="headline-2">Headline 2</Typography>
 *       <Typography type="headline-3">Headline 3</Typography>
 *       <Typography type="headline-4">Headline 4</Typography>
 *       <Typography type="headline-5">Headline 5</Typography>
 *       <Typography type="headline-6">Headline 6</Typography>
 *       <Typography type="subtitle-1">Subtitle 1</Typography>
 *       <Typography type="subtitle-2">Subtitle 2</Typography>
 *       <Typography type="body-1">Body 1</Typography>
 *       <Typography type="body-2">Body 2</Typography>
 *       <Typography type="caption" component="h5">
 *         Caption text
 *       </Typography>
 *       <Typography type="overline" component="h5">
 *         Overline text
 *       </Typography>
 *       <Typography type="button" component="h5">
 *         Button text
 *       </Typography>
 *     </>
 *   );
 * }
 * ```
 *
 * @remarks \@since REPLACE_VERSION Added an example to the documentation.
 */
export const Typography = forwardRef<TypographyHTMLElement, TypographyProps>(
  function Typography(
    {
      className: propClassName,
      children,
      component = null,
      align,
      color,
      decoration,
      fontStyle,
      margin,
      transform,
      type = "body-1",
      weight,
      ...props
    },
    ref
  ) {
    const className = typographyClassName(
      {
        align,
        color,
        decoration,
        fontStyle,
        margin,
        transform,
        type,
        weight,
      },
      propClassName
    );
    if (typeof children === "function") {
      return (children as TypographyRenderFunction)({ className });
    }

    const Component = getComponent(component, type);
    return (
      <Component {...props} ref={ref} className={className}>
        {children}
      </Component>
    );
  }
);
