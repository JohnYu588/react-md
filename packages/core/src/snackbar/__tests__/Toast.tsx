import { describe, expect, it, jest } from "@jest/globals";
import type { ReactElement } from "react";
import { createRef } from "react";
import type { RenderResult } from "../../test-utils/index.js";
import {
  rmdRender as baseRender,
  screen,
  userEvent,
} from "../../test-utils/index.js";

import { FontIcon } from "../../icon/FontIcon.js";
import { Typography } from "../../typography/Typography.js";
import { Toast } from "../Toast.js";
import { CurrentToastActionsProvider } from "../useCurrentToastActions.js";

const render = (ui: ReactElement): RenderResult =>
  baseRender(ui, {
    wrapper: ({ children }) => (
      <CurrentToastActionsProvider
        value={{
          clearTimer: jest.fn(),
          pauseRemoveTimeout: jest.fn(),
          removeToast: jest.fn(),
          resumeRemoveTimeout: jest.fn(),
          startRemoveTimeout: jest.fn(),
        }}
      >
        {children}
      </CurrentToastActionsProvider>
    ),
  });

describe("Toast", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      "data-testid": "toast",
      ref,
      visible: true,
    } as const;
    const { rerender } = render(<Toast {...props} />);
    const toast = screen.getByTestId("toast");
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(toast);
    expect(toast).toMatchSnapshot();

    rerender(
      <Toast
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(toast).toMatchSnapshot();
  });

  it("should render any valid element within the ToastActionButton", async () => {
    const user = userEvent.setup();
    const props = {
      children: <Typography>Hello, world!</Typography>,
      visible: true,
    } as const;
    const { rerender } = render(<Toast {...props} />);

    expect(() => screen.getByRole("button")).toThrow();

    rerender(<Toast {...props} action={<span>Action</span>} />);
    const action = screen.getByRole("button", { name: "Action" });
    expect(action).toBeInTheDocument();
    expect(action).toMatchSnapshot();

    const onClick = jest.fn();
    rerender(
      <Toast
        {...props}
        action={{
          onClick,
          children: <FontIcon>favorite</FontIcon>,
          buttonType: "icon",
          "aria-label": "New Action",
        }}
      />
    );

    const newAction = screen.getByRole("button", { name: "New Action" });
    expect(newAction).toMatchSnapshot();

    await user.click(newAction);
    expect(onClick).toHaveBeenCalled();
  });

  it("should not be in the dom while not visible", () => {
    const { getByTestId } = render(
      <Toast data-testid="toast" visible={false}>
        Hello
      </Toast>
    );

    expect(() => getByTestId("toast")).toThrow();
  });
});
