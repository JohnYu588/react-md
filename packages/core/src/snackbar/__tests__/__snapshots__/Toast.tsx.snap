// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<div
  class="rmd-toast rmd-toast--surface rmd-scale-transition--enter rmd-scale-transition--enter-active"
  data-testid="toast"
  id="toast-:r0:"
>
  <div
    class="rmd-toast-content"
  />
</div>
`;

exports[`Toast should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<div
  class="rmd-toast rmd-toast--surface custom-class-name rmd-scale-transition--enter rmd-scale-transition--enter-active"
  data-testid="toast"
  id="toast-:r0:"
  style="color: white;"
>
  <div
    class="rmd-toast-content"
  />
</div>
`;

exports[`Toast should render any valid element within the ToastActionButton 1`] = `
<button
  class="rmd-button rmd-button--text rmd-button--text-secondary rmd-toast-action"
  type="button"
>
  <span>
    Action
  </span>
</button>
`;

exports[`Toast should render any valid element within the ToastActionButton 2`] = `
<button
  aria-label="New Action"
  class="rmd-button rmd-button--icon rmd-button--text-secondary rmd-toast-action"
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </span>
</button>
`;
